### DISCUSSION TITLE
NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer

### DISCUSSION THREAD

**Discussion by vitsin** [jchambers/java-otp#26] [Q&A] [answered]
The following code as is from test ExampleApp.java returns error:
```
Exception in thread "main" java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer;
	at com.eatthepath.otp.HmacOneTimePasswordGenerator.generateOneTimePassword(HmacOneTimePasswordGenerator.java:140)
	at com.eatthepath.otp.TimeBasedOneTimePasswordGenerator.generateOneTimePassword(TimeBasedOneTimePasswordGenerator.java:143)
	at com.eatthepath.otp.TimeBasedOneTimePasswordGenerator.generateOneTimePasswordString(TimeBasedOneTimePasswordGenerator.java:175)
	at com.eatthepath.otp.TimeBasedOneTimePasswordGenerator.generateOneTimePasswordString(TimeBasedOneTimePasswordGenerator.java:160)
	at simple.ExampleApp.main(ExampleApp.java:30)
```

```java
package simple;

import javax.crypto.KeyGenerator;

import com.eatthepath.otp.TimeBasedOneTimePasswordGenerator;

import java.security.InvalidKeyException;
import java.security.Key;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;

public class ExampleApp {
    public static void main(final String[] args) throws NoSuchAlgorithmException, InvalidKeyException {
        final TimeBasedOneTimePasswordGenerator totp = new TimeBasedOneTimePasswordGenerator();

        final Key key;
        {
            final KeyGenerator keyGenerator = KeyGenerator.getInstance(totp.getAlgorithm());

            // Key length should match the length of the HMAC output (160 bits for SHA-1, 256 bits
            // for SHA-256, and 512 bits for SHA-512).
            keyGenerator.init(160);

            key = keyGenerator.generateKey();
        }

        final Instant now = Instant.now();
        final Instant later = now.plus(totp.getTimeStep());

        System.out.println("Current password: " + totp.generateOneTimePasswordString(key, now));
        System.out.println("Future password:  " + totp.generateOneTimePasswordString(key, later));
    }
}
```

**Comment by vitsin** [comment]
Running on Java 1.8.0 u281

**Reply by jchambers** [reply]
How are you determining the version? Is it possible that the version of Java that's running your application is different from the version that's building it? What happens if you add this somewhere before the calls to `generateOneTimePasswordString`?

```java
System.out.println("java.version: " + System.getProperty("java.version"));
```

**Reply by vitsin** [reply]
java.version: 1.8.0_281, from Oracle

**Reply by vitsin** [reply]
I took the java-otp-0.3.0.jar from https://github.com/jchambers/java-otp/releases.

Found also this interesting idea: if the jar is built using java9 and I run it with java8 than:

> https://stackoverflow.com/questions/48693695/java-nio-buffer-not-loading-clear-method-on-runtime

**Reply by Dhananjay-J-P** [reply]
I am also getting the same error for 0.3.0 with Java 1.8.x @jchambers

**Comment by vitsin** [comment]
ok, found the issue! 
java-otp-0.3.0.jar from https://github.com/jchambers/java-otp/releases is compiled with JDK 9 and is not working with JDK8 (see original exception).
Once I've recompiled your source code with JDK 8 - all works fine.

**Reply by jchambers** [reply]
Huh… interesting! I'm surprised that's the case; do you have the some problem if you use Maven or Gradle to include java-otp as a dependency in your project?

**Reply by jchambers** [reply]
One more clarification here: can you explain _exactly_ how you were running things before? I have a hunch that the main Java runtime jar wasn't on your classpath before, and that may have been the root problem here. I don't mean to point fingers, to be clear; it's just that this behavior seems really weird, and I want to make sure I'm not uploading bad artifacts somewhere.

**Reply by vitsin** [reply]
when I add java-otp-0.3.0.jar from https://github.com/jchambers/java-otp/releases directly into eclipse libraries or alternatively using pom.xml with
		```<dependency>
			<groupId>com.eatthepath</groupId>
			<artifactId>java-otp</artifactId>
			<version>0.3.0</version>
		</dependency>```
in both cases, when using JDK 8 u281, I am getting runtime exception "java.lang.NoSuchMethodError: java.nio.ByteBuffer.clear()Ljava/nio/ByteBuffer".

To simplify: 
when I add java-otp-0.3.0.jar directly into eclipse - runtime exception. 
Once I add compiled jar from your source on my dev machine exact same way as java-otp-0.3.0.jar - all works fine.

**Comment by jchambers** [comment, marked as answer]
Folks, it sounds like there's a legit problem here. Somebody else has already opened an issue, so let's move the discussion to https://github.com/jchambers/java-otp/issues/27.

**Reply by jchambers** [reply]
Folks, this should be fixed in #28. It appears this is a fairly widespread issue when using `ByteBuffer` across the Java 8/9 boundary: https://www.morling.dev/blog/bytebuffer-and-the-dreaded-nosuchmethoderror/

**Reply by jchambers** [reply]
#28 has been merged and version 0.3.1 has been released (though it will likely take a couple of hours to make its way to Maven Central). Please let me know if problems persist after updating.
